### ISSUE TITLE
Async validator inconsistencies

### ISSUE THREAD

**Issue by benmccallum** [jeffhandley/strickland#17]
Repro steps:
1. Run react demo.
2. Enter 'benmccallum' into username field (don't leave focus)
3. Pause for a second to let async validation fire and notice JSON output: it shows `value: 'benmccallum'`.
4. Add a space to username field (don't leave focus)
5. Notice JSON output is now missing `username` results, **should it empty here?** (Issue A)
6. Remove space previously added to username field (don't leave focus).
7. Notice async validation fires again, with same value **should it re-run here?** (Issue B)

Issue C:
Now, reload the page and do the same but after step 4, **do** leave focus. This causes the async validation to fire with value 'benmccallum ', **so it runs on blur when space is added after but not on change**. Must be something to do with change detection. ~At first I thought this was something to do with the trimming logic, but the username field doesn't have trim.~ Does have trim, just needs adding to `onchange` as well as `onblur` event. ✔️ 

(Note: I'm working through some other issues on the same field on the vue demo so don't go looking at it for inspiration. And I actually use trim on it, which is probably more appropriate anyway.)

**Comment by jeffhandley** [comment]
I think this is happening because the username field has an additional validator applied to it that ensures there are no spaces in the value.  So with spaces added in, the value becomes invalid in synchronous validation, and therefore async validation would not fire.

In the React demos, I was separating `form` from `parsedForm` in state so that trim could be applied, put into the `parsedForm` state, and then validation occurs against `parsedForm`.  This separation allowed validation to run against what the trimmed result would be, meaning that typing trailing or leading spaces into a trimmed field would not affect the value to validate, and even synchronous validation would not need to run again.

Do you think some of this handling could be causing the behavior you're seeing?

**Comment by benmccallum** [comment]
I found an inconsistency which threw me off for a bit. Trim was applied to `change` but not `blur`, I've made it consistent and will keep digging.

**Comment by benmccallum** [comment]
Thanks for the heads up, it certainly could be, I'll take a look!

**Comment by benmccallum** [comment]
Another issue now that trim is enabled onblur. Noting here to come back to.

Repro:
1. Open react demo (component state or redux)
2. Enter `benmccallum` into username field, allow to validate (don't leave).
3. Add space (don't leave) (no validation occurs, issue above).
4. Now leave by tabbing out.

Result: 
* Validation message is stuck to always 'Username: checking availability of "benmccallum"'.
* Confirmed that it's the automatic race condition handling.

**Comment by benmccallum** [comment]
Issues A, B, C and D I found above should all be covered by PR: https://github.com/jeffhandley/strickland/issues/20

**Comment by benmccallum** [comment]
Closing. All issues now addressed in merged PRs.
